### PR TITLE
Make HMR work when locals are changed, but with test repo to view it

### PR DESCRIPTION
**Pull request by KevinCChan**
**What kind of change does this PR introduce?**
Bugfix for https://github.com/webpack/style-loader/issues/320

**Did you add tests for your changes?**
No. There is not a direct way to test this change due to the many moving parts.

I created a repo at https://github.com/KevinCChan/css-hmr-bugfix to demo the change. All it does is open run a webserver on 127.0.0.1:8080 with a colored box to demo prefix and postfix behavior.

**If relevant, did you update the README?**
Yes

**Summary**
This is a bug fix for the crashing behavior introduced by https://github.com/webpack-contrib/style-loader/pull/298. There was a similar pull request, and we came to the same conclusion. I am doing my part and creating a test repository to acceptance test the changes so the fix can be integrated. I updated documentation as well.

**Does this PR introduce a breaking change?**
No, I made this a toggle option specifically to avoid introducing a breaking change.

**Other information**
Please reach to me if there are any issues!

### CODE REVIEW

**Comment by KevinCChan** [comment]
@evilebottnawi

There is no automatic test that can be written. I styled this PR similar to https://github.com/webpack-contrib/style-loader/pull/298 which was accepted without a test and provided a test repo to view the change.

**Comment by evilebottnawi** [comment]
@KevinCChan don't think we need new option, we should fix it without new options

**Comment by evilebottnawi** [comment]
Also this problem more related to dev server, not loader

**Comment by KevinCChan** [comment]
@evilebottnawi

If we want to fix without new option, then I feel the correct course of action is to revert https://github.com/webpack-contrib/style-loader/pull/298.

https://github.com/webpack-contrib/style-loader/pull/298 specifically targets webpack-dev-server (a specific implementation of express/webpack-dev-middleware) at the cost of regressing functionality for every other implementation of express/webpack-dev-middleware.

I can look into alternatives more when I get back home tonight.

**Comment by KevinCChan** [comment]
@evilebottnawi 

I did some digging around source code and playing with the repos and debuggers. This is what I found.

https://github.com/webpack-contrib/style-loader/pull/298 specifically targeted webpack-dev-server (WDS) with the hot option which meant that webpack-dev-server's included full reloaders would kick in.

If someone instead used WDS's hotOnly option in lieu of the hot option (which is documented [here](https://webpack.js.org/configuration/dev-server/#devserver-hotonly)), WDS behaves like a generic Express/webpack-dev-middleware/webpack-hot-middleware server configuration, much like my own and many others. You will see that https://github.com/webpack-contrib/style-loader/pull/298 actually fails in that specific configuration because the underlying assumption (of an error handler) no longer exists!

This makes the option toggle more relevant since it's impossible to determine at runtime whether or not an error handler has been provided to fully reload the page. Having the option at least allows users to configure the behavior of style-loader to match their given environment.

I recommend revisiting the test repo used as an acceptance test for https://github.com/webpack-contrib/style-loader/pull/298 (found [here](https://github.com/lydell/css-hmr-bug/tree/fix)) and editting the webpack_config from hot -> hotOnly and observe that issue https://github.com/webpack-contrib/style-loader/issues/320 existed then too.

**Comment by sasha240100** [comment]
@KevinCChan @evilebottnawi Any updates to it?

**Comment by lydell** [comment]
I think this is a good workaround, though I'm no expert (as evidenced by my previous PR, which btw, didn't intentionally target just webpack-dev-server but accidentally did so). :+1:

The "real", but much harder, solution is to add support in webpack's hot reloading "framework" for conditionally accepting hot updates depending on their contents, I think. Pinging @sokra just in case you have the time to share your wisdom, but I also understand if you're swamped with notifications and other stuff to do.

**Comment by lydell** [comment]
Thinking a bit more about this, I thought that webpack-dev-server was just a convenience wrapper around webpack-dev-middleware, so to me it’s surprising that they behave differently. So maybe we should also make them behave the same? Not sure what the “correct” behavior is though.

**Comment by KevinCChan** [comment]
@evilebottnawi @lydell 

This [issue](https://github.com/webpack/webpack/issues/418) led to the adoption of `hot` and `hotOnly` options in WDS.

TL;DR:
Sokra (webpack maintainer) and gaearon (react-hot-loader maintainer) discussed the exact issue that https://github.com/webpack-contrib/style-loader/pull/298 addressed about auto-reloading. It was decided that two implementations were necessary.

If the direction that style-loader wants to follow is one that is perfectly compatible with webpack then a new option is unavoidable due to the precedence set above in https://github.com/webpack/webpack/issues/418 and https://github.com/webpack/webpack/commit/e43a76a65e5511ebf58c9bc4983530869a48c921.

If style-loader only wants to support the `hot` option, then I will close this pull request.

**Comment by KevinCChan** [comment]
@evilebottnawi No problem! I have been busy as of late so I did not really have time to respond. Let me see what is possible on the css-loader level
